File: lib_base/src/main/java/com/android/base/app/ApplicationDelegate.java
package com.android.base.app;

import android.app.Application;
import android.content.Context;
import android.content.IntentFilter;
import android.content.res.Configuration;
import android.net.ConnectivityManager;

import com.android.base.receiver.NetStateReceiver;
import com.android.base.utils.BaseUtils;
import com.blankj.utilcode.util.AppUtils;
import com.blankj.utilcode.util.Utils;

import java.util.concurrent.atomic.AtomicBoolean;

import io.reactivex.Flowable;
import io.reactivex.processors.BehaviorProcessor;
import timber.log.Timber;

/**
 * @author Ztiany
 * Email: ztiany3@gmail.com
 * Date : 2018-10-12 18:19
 */
@SuppressWarnings("WeakerAccess,unused")
public final class ApplicationDelegate {

    private Application mApplication;

    private CrashHandler mCrashHandler;
    private BehaviorProcessor<Boolean> mAppStatus;

    private AtomicBoolean mOnCreateCalled = new AtomicBoolean(false);
    private AtomicBoolean mOnAttachBaseCalled = new AtomicBoolean(false);

    ApplicationDelegate() {
    }

    public void attachBaseContext(@SuppressWarnings("unused") Context base) {
        if (!mOnAttachBaseCalled.compareAndSet(false, true)) {
            throw new IllegalStateException("Can only be called once");
        }
        //no op
    }

    public void onCreate(Application application) {
        if (!mOnCreateCalled.compareAndSet(false, true)) {
            throw new IllegalStateException("Can only be called once");
        }
        mApplication = application;
        //工具类初始化
        BaseUtils.init(application);
        //异常日志记录
        mCrashHandler = CrashHandler.register(application);
        //网络状态
        listenNetworkState();
        //App前台后台
        listenActivityLifecycleCallbacks();
    }

    public void onTerminate() {
        //no op
    }

    public void onConfigurationChanged(Configuration newConfig) {
        //no op
    }

    public void onTrimMemory(int level) {
        //no op
    }

    public void onLowMemory() {
        //no op
    }

    private void listenNetworkState() {
        NetStateReceiver netStateReceiver = new NetStateReceiver();
        mApplication.registerReceiver(netStateReceiver, new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION));
    }

    private void listenActivityLifecycleCallbacks() {
        mAppStatus = BehaviorProcessor.create();
        AppUtils.registerAppStatusChangedListener(this, new Utils.OnAppStatusChangedListener() {
            @Override
            public void onForeground() {
                Timber.d("app进入前台");
                mAppStatus.onNext(true);
            }

            @Override
            public void onBackground() {
                Timber.d("app进入后台");
                mAppStatus.onNext(false);
            }
        });
    }

    /**
     * 获取可观察的 app 生命周期
     */
    Flowable<Boolean> appAppState() {
        return mAppStatus;
    }

    void setCrashProcessor(Sword.CrashProcessor crashProcessor) {
        if (mCrashHandler != null) {
            mCrashHandler.setCrashProcessor(crashProcessor);
        }
    }

    Application getApplication() {
        return mApplication;
    }

}

File: lib_base/src/main/java/com/android/base/app/ApplicationDelegate.kt
package com.android.base.app

import android.app.Application
import android.content.Context
import android.content.IntentFilter
import android.content.res.Configuration
import android.net.ConnectivityManager
import com.android.base.app.utils.CrashHandler
import com.android.base.receiver.NetStateReceiver
import com.android.base.utils.BaseUtils
import com.blankj.utilcode.util.AppUtils
import com.blankj.utilcode.util.Utils.OnAppStatusChangedListener
import io.reactivex.processors.BehaviorProcessor
import timber.log.Timber
import java.util.concurrent.atomic.AtomicBoolean

/**
 * @author Ztiany
 * Email: ztiany3@gmail.com
 * Date : 2018-10-12 18:19
 */
internal class ApplicationDelegate internal constructor(private val androidComponentLifecycleInjector: AndroidComponentLifecycleInjector) {

    lateinit var application: Application

    private lateinit var crashHandler: CrashHandler

    /** 获取可观察的 app 生命周期  */
    val appStatus: BehaviorProcessor<Boolean> = BehaviorProcessor.create()

    private val onCreateCalled = AtomicBoolean(false)
    private val onAttachBaseCalled = AtomicBoolean(false)

    fun attachBaseContext(base: Context) {
        check(onAttachBaseCalled.compareAndSet(false, true)) { "Can only be called once" }
    }

    fun onCreate(application: Application) {
        check(onCreateCalled.compareAndSet(false, true)) { "Can only be called once" }
        this.application = application
        //工具类初始化
        BaseUtils.init(application)
        //异常日志记录
        crashHandler = CrashHandler.register(application)
        //网络状态
        application.registerReceiver(NetStateReceiver(), IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION))
        //App前台后台
        listenActivityLifecycleCallbacks()
        //声明周期回调
        application.registerActivityLifecycleCallbacks(androidComponentLifecycleInjector)
    }

    fun onTerminate() {}

    fun onConfigurationChanged(newConfig: Configuration) {}

    fun onTrimMemory(level: Int) {}

    fun onLowMemory() {}

    private fun listenActivityLifecycleCallbacks() {
        AppUtils.registerAppStatusChangedListener(this, object : OnAppStatusChangedListener {
            override fun onForeground() {
                Timber.d("app进入前台")
                appStatus.onNext(true)
            }

            override fun onBackground() {
                Timber.d("app进入后台")
                appStatus.onNext(false)
            }
        })
    }

    internal fun setCrashProcessor(crashProcessor: CrashProcessor) {
        crashHandler.setCrashProcessor(crashProcessor)
    }

}